In [1]:
import tensorflow as tf

In [2]:
print(tf.__version__)

2.9.1


In [3]:
import numpy as np
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

In [4]:
model = Sequential([Dense(units=1, input_shape=[1])])
model.compile(optimizer='sgd', loss='mean_squared_error')

xs = np.array([-1.0, 0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float)

model.fit(xs, ys, epochs=500)

print(model.predict([10.0]))

Epoch 1/500
1/1 [==============================] - 2s 2s/step - loss: 22.4876
Epoch 2/500
1/1 [==============================] - 0s 5ms/step - loss: 17.9707
Epoch 3/500
1/1 [==============================] - 0s 2ms/step - loss: 14.4112
Epoch 4/500
1/1 [==============================] - 0s 3ms/step - loss: 11.6052
Epoch 5/500
1/1 [==============================] - 0s 4ms/step - loss: 9.3921
Epoch 6/500
1/1 [==============================] - 0s 2ms/step - loss: 7.6455
Epoch 7/500
1/1 [==============================] - 0s 2ms/step - loss: 6.2661
Epoch 8/500
1/1 [==============================] - 0s 3ms/step - loss: 5.1758
Epoch 9/500
1/1 [==============================] - 0s 2ms/step - loss: 4.3129
Epoch 10/500
1/1 [==============================] - 0s 2ms/step - loss: 3.6290
Epoch 11/500
1/1 [==============================] - 0s 2ms/step - loss: 3.0862
Epoch 12/500
1/1 [==============================] - 0s 2ms/step - loss: 2.6543
Epoch 13/500
1/1 [==============================] - 0s 2ms

In [6]:
dense = Dense(units=1, input_shape=[1]) # Dense 층을 사용함 1개의 뉴런임
model = Sequential([dense]) # Sequential 클래스로 층을 정의함(여기서는 하나의 층임)
model.compile(optimizer='sgd', loss='mean_squared_error') # 손실함수(확률적 경사 하강법)

xs = np.array([-1.0, 0.0, 1.0, 2.0, 3.0, 4.0], dtype=float) # 데이터 1
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float) # 데이터 2

model.fit(xs, ys, epochs=500) # 500번 훈련

print(model.predict([10.0])) # (X, 10을 넣었을 때) 예측실행
print("신경망이 학습한 것: {}".format(dense.get_weights())) # Y = 2X - 1과 비슷한 값이 나옴
# Y = 1.998174X -0.9943387

Epoch 1/500
1/1 [==============================] - 0s 143ms/step - loss: 0.5996
Epoch 2/500
1/1 [==============================] - 0s 3ms/step - loss: 0.5848
Epoch 3/500
1/1 [==============================] - 0s 2ms/step - loss: 0.5708
Epoch 4/500
1/1 [==============================] - 0s 2ms/step - loss: 0.5576
Epoch 5/500
1/1 [==============================] - 0s 3ms/step - loss: 0.5449
Epoch 6/500
1/1 [==============================] - 0s 2ms/step - loss: 0.5328
Epoch 7/500
1/1 [==============================] - 0s 2ms/step - loss: 0.5211
Epoch 8/500
1/1 [==============================] - 0s 2ms/step - loss: 0.5098
Epoch 9/500
1/1 [==============================] - 0s 2ms/step - loss: 0.4988
Epoch 10/500
1/1 [==============================] - 0s 2ms/step - loss: 0.4882
Epoch 11/500
1/1 [==============================] - 0s 2ms/step - loss: 0.4779
Epoch 12/500
1/1 [==============================] - 0s 3ms/step - loss: 0.4679
Epoch 13/500
1/1 [==============================] - 0s 2ms/

In [8]:
data = tf.keras.datasets.fashion_mnist # 데이터셋 불러오기

(training_images, training_labels), (test_images, test_labels) = data.load_data()
# label은 0~9
# training 이미지는 28 * 28 6만개, test 1만개

training_images = training_images / 255.0 # 255로 나누어 정규화
test_images = test_images / 255.0
 
model = tf.keras.models.Sequential([tf.keras.layers.Flatten(input_shape=(28, 28)),
                                   tf.keras.layers.Dense(128, activation = tf.nn.relu),
                                   tf.keras.layers.Dense(10, activation=tf.nn.softmax)
                                   ])

model.compile(optimizer='adam', # sgd 옵티마이저가 진화한 adam 옵티마이저
              loss='sparse_categorical_crossentropy', # 희소한 범주형 크로스 엔트로피 ( 카테고리 선택)
              metrics=['accuracy']) # 정확도 리포트

model.fit(training_images, training_labels, epochs = 5) # 5번 훈련

Epoch 1/5
1875/1875 [==============================] - 85s 45ms/step - loss: 0.5018 - accuracy: 0.8246
Epoch 2/5
1875/1875 [==============================] - 75s 40ms/step - loss: 0.3764 - accuracy: 0.8635
Epoch 3/5
1875/1875 [==============================] - 84s 45ms/step - loss: 0.3383 - accuracy: 0.8773
Epoch 4/5
1875/1875 [==============================] - 83s 44ms/step - loss: 0.3131 - accuracy: 0.8859
Epoch 5/5
1875/1875 [==============================] - 82s 44ms/step - loss: 0.2967 - accuracy: 0.8905


In [9]:
model.evaluate(test_images, test_labels) # 87.65%의 정확도

313/313 [==============================] - 15s 43ms/step - loss: 0.3444 - accuracy: 0.8765


[0.3444223999977112, 0.8765000104904175]

In [10]:
classifications = model.predict(test_images)
print(classifications[0])
print(test_labels[0]) # 92.3%의 확률로 9번 레이블

313/313 [==============================] - 13s 43ms/step
[2.8476541e-06 3.3549625e-08 4.5595368e-07 2.8592163e-08 5.5549281e-06
 2.9404536e-03 4.4913850e-06 7.3552348e-02 2.9288316e-05 9.2346442e-01]
9


# 과대적합
앞에서는 5번의 에폭 동안만 훈련했습니다. 즉, 뉴련을 랜덤하게 초기화한 후에 훈련 세트를 모델에 전달해 그 결과를 레이블과 비교하고, 손실 함수로 성능을 측정하고, 그 다음 옵티마이저를 업데이트하는 전체 훈련 과정을 5번 수행했습니다. 이로써 얻은 결과는 꽤 훌륭했습니다. 훈련 세트에서 89%의 정확도, 테스트 세트에서 87%의 정확도를 달성했습니다. 더 오래 훈련하면 어떨까요??  

에폭 수를 5에서 50으로 늘려 훈련해보죠. 훈련 세트에서 얻은 정확도는 다음과 같습니다. 기존보다 훨씬 높은 96.27%의 정확도를 달성했습니다. 다음과 같이 테스트 세트의 정확도는 88.6% 입니다.  
 
훈련세트의 정확도는 크게 높아졌지만 테스트 세트에 대한 정확도는 조금만 향상되었습니다. 오래 훈련했기 때문에 더 높은 성능이 나오긴 했지만 항상 이러한 결과가 나오지는 않습니다. 이 신경망은 훈련 세트에서 훨씬 높은 성능을 달성했지만 월등하게 더 나은 모델은 아닙니다. 사실 정확도 차이가 더 커졌으므로 모델이 훈련 세트에 특화되었음을 보여줍니다. 이를 과대적합이라 부릅니다. 신경망을 만들 떄 주의해야 할 문제이므로 나중에 이를 피할 수 있는 여러가지 기법을 소개합니다.

In [12]:
class myCallback(tf.keras.callbacks.Callback):
    def on_epoch_end(self, epoch, logs={}):
        if(logs.get('accuracy')>0.9):
            print("\n정확도 90%에 도달하여 훈련을 멈춥니다!")
            self.model.stop_training = True
            
callbacks = myCallback()
mnist = tf.keras.datasets.fashion_mnist

(training_images, training_labels), (test_images, test_labels) = mnist.load_data()


training_images = training_images / 255.0
test_images = test_images / 255.0
 
model = tf.keras.models.Sequential([tf.keras.layers.Flatten(input_shape=(28, 28)),
                                   tf.keras.layers.Dense(128, activation = tf.nn.relu),
                                   tf.keras.layers.Dense(10, activation=tf.nn.softmax)
                                   ])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(training_images, training_labels, epochs = 50, callbacks=[callbacks])

Epoch 1/50
1875/1875 [==============================] - 82s 44ms/step - loss: 0.5058 - accuracy: 0.8225
Epoch 2/50
1875/1875 [==============================] - 79s 42ms/step - loss: 0.3837 - accuracy: 0.8632
Epoch 3/50
1875/1875 [==============================] - 82s 43ms/step - loss: 0.3433 - accuracy: 0.8753
Epoch 4/50
1875/1875 [==============================] - 81s 43ms/step - loss: 0.3165 - accuracy: 0.8843
Epoch 5/50
1875/1875 [==============================] - 81s 43ms/step - loss: 0.2974 - accuracy: 0.8903
Epoch 6/50
1875/1875 [==============================] - 78s 42ms/step - loss: 0.2825 - accuracy: 0.8962
Epoch 7/50
1875/1875 [==============================] - 80s 42ms/step - loss: 0.2687 - accuracy: 0.8999
Epoch 8/50
1874/1875 [============================>.] - ETA: 0s - loss: 0.2570 - accuracy: 0.9050
정확도 90%에 도달하여 훈련을 멈춥니다!
1875/1875 [==============================] - 80s 43ms/step - loss: 0.2570 - accuracy: 0.9050
